src/org/javarosa/xform/parse/XFormParser.java

```diff
@@ -388,55 +387,7 @@ public static Document getXMLDocument(Reader reader, CacheTable<String> stringCa
             Std.out.println("Error closing reader");
             Std.printStack(e);
         }
-
-        //For escaped unicode strings we end up with a looooot of cruft,
-        //so we really want to go through and convert the kxml parsed
-        //text (which have lots of characters each as their own string)
-        //into one single string
-        Stack<Element> q = new Stack<>();
-
-        q.push(doc.getRootElement());
-        while(!q.isEmpty()) {
-            Element e = q.pop();
-            boolean[] toRemove = new boolean[e.getChildCount()*2];
-            String accumulate = "";
-            for(int i = 0 ; i < e.getChildCount(); ++i ){
-                int type = e.getType(i);
-                if(type == Element.TEXT) {
-                    String text = e.getText(i);
-                    accumulate += text;
-                    toRemove[i] = true;
-                } else {
-                    if(type ==Element.ELEMENT) {
-                        q.add(e.getElement(i));
-                    }
-                    String accumulatedString = accumulate.trim();
-                    if(accumulatedString.length() != 0) {
-                        if(stringCache == null) {
-                            e.addChild(i, Element.TEXT, accumulate);
-                        } else {
-                            e.addChild(i, Element.TEXT, stringCache.intern(accumulate));
-                        }
-                        accumulate = "";
-                        ++i;
-                    } else {
-                        accumulate = "";
-                    }
-                }
-            }
-            if(accumulate.trim().length() != 0) {
-                if(stringCache == null) {
-                    e.addChild(Element.TEXT, accumulate);
-                } else {
-                    e.addChild(Element.TEXT, stringCache.intern(accumulate));
-                }
-            }
-            for(int i = e.getChildCount() - 1; i >= 0 ; i-- ){
-                if(toRemove[i]) {
-                    e.removeChild(i);
-                }
-            }
-        }
+        XmlTextConsolidator.consolidateText(stringCache, doc.getRootElement());
```

Comment:
@dcbriccetti I've been wondering why do we need `stringCache` and `InterningKXmlParser` but I've never had a chance/time to investigate this properly. 
Do you know what is this for and under what condition is it used?
I'm just hoping that this may be coming from "the dark J2ME" days and we could get rid of this at some point :)

Reply:
I wondered the same, but didn’t look into it.